Enhance DAC to allow getting AssemblyLoadContext for objects

janvorli opened the pull request:
This change adds support to DAC for getting AssemblyLoadContext
for a MethodTable. The returned AsseblyLoadContext is the context
into which the related type was loaded.

jkotas commented on src/coreclr/src/debug/daccess/dacimpl.h:
```diff
@@ -840,7 +840,8 @@ class ClrDataAccess
       public ISOSDacInterface6,
-      public ISOSDacInterface7
+      public ISOSDacInterface7,
+      public ISOSDacInterface8
```
FYI @davmason is adding this in https://github.com/dotnet/runtime/pull/37853 as well.

jkotas commented on src/coreclr/src/inc/sospriv.idl:
```diff
@@ -390,3 +390,14 @@ interface ISOSDacInterface7 : IUnknown
+]
+interface ISOSDacInterface8 : IUnknown
+{
+    HRESULT GetAssemblyLoaderAllocatorData(CLRDATA_ADDRESS loaderAllocator, struct DacpAssemblyLoaderAllocatorData* pAssemblyLoaderAllocatorData);
```
Where do you get the address of the loader allocator or the binder from? Do we also need APIs to e.g. get loader allocator from MethodTable or MethodDesc?

janvorli replied:
We already have that, I have added that quite some time ago.
```
interface ISOSDacInterface6 : IUnknown
{
    HRESULT GetMethodTableCollectibleData(CLRDATA_ADDRESS mt, struct DacpMethodTableCollectibleData *data);
};
```
I also have the SOS change tested and ready to get in.

jkotas replied:
`GetMethodTableCollectibleData` returns `LoaderAllocatorObjectHandle`. How do you get from `LoaderAllocatorObjectHandle` to `loaderAllocator` ?

>  SOS change

It may be useful to add link to the SOS change for context.

janvorli replied:
See https://github.com/dotnet/runtime/issues/11157#issuecomment-642288660 for the steps to get from an object to the AssemblyLoadContext. There were two steps that couldn't be done via DAC before, those are the ones I'm adding here.

As for the SOS change, I was planning to create the PR after this one gets in.

mikem8361 replied:
You may want to write some of the SOS code to test these new DAC APIs to test the binder code added to the DAC has been properly "DACized" (dac_cast<T> macros, etc).

jkotas commented on src/coreclr/src/inc/sospriv.idl:
```diff
@@ -390,3 +390,14 @@ interface ISOSDacInterface7 : IUnknown
+interface ISOSDacInterface8 : IUnknown
+{
+    HRESULT GetAssemblyLoaderAllocatorData(CLRDATA_ADDRESS loaderAllocator, struct DacpAssemblyLoaderAllocatorData* pAssemblyLoaderAllocatorData);
+    HRESULT GetCLRPrivBinderAssemblyLoadContextData(CLRDATA_ADDRESS binder, struct DacpCLRPrivBinderAssemblyLoadContextData* pBinderData);
```
`CLRPrivBinder` is a poor internal name. I would use the public facing name for this, ie AssemblyLoadContext.

janvorli replied:
Sounds reasonable.

janvorli commented:
Hmm, it seems that all Linux builds are broken due to a missing header. I am looking into it.

jkotas commented on src/coreclr/src/inc/dacprivate.h:
```diff
@@ -1031,6 +1031,42 @@ struct MSLAYOUT DacpJitCodeHeapInfo
 
+struct MSLAYOUT DacpAssemblyLoaderAllocatorData
+{
+    CLRDATA_ADDRESS BinderToRelease;
```
Use more self-describing name, e.g. `OwningAssemblyLoadContext` ?

jkotas commented:
It looks good to me otherwise. I would like to see somebody from the diagnostic team to signoff on this as well.

tommcdon commented:
> It looks good to me otherwise. I would like to see somebody from the diagnostic team to signoff on this as well.

@mikem8361 @sdmaclea can you take a look?

janvorli commented:
I am going to redo this change to add just one new method to get pointer to managed AssemblyLoadContext for a given MethodTable. The current way worked only for collectible AssemblyLoadContext and I have realized that I can make it work for both collectible and non-collectible. 
My original change was trying to expose separate data structures for a subset of members of specific native structures (subset == 1 field in this case) and use it as two of multiple steps in SOS to get to the AssemblyLoadContext. But it seems that the ISOSDacInterface methods don't follow such philosophy in general, so I can just expose a method that internally does those steps on the DAC side (basically a simple pointer chain walk) and returns just the resulting AssemblyLoadContext)

janvorli commented:
@jkotas, @mikem8361 can you please take a look again?  I've made substantial changes to you'll probably want to review it from scratch. I've redone the implementation to support 
dumping AssemblyLoadContext even for the cases when it is the default one or a non-collectible one. I have realized that the previous way of exposing some partial data structures is not going to work well here. So I've done it differently by adding a higher level DAC method to get AssemblyLoadContext for a MethodTable and implement all the steps needed to get from MethodTable to the ALC there. The need to access other data structures to be able to get ALC for
non-collectible types turned out into a DACization waterfall and I even had to add special handling for classes in namespaces that the DAC would not be able to handle and even update the DacTableGen to support that (symbols of classes inside namespaces are mangled in a special way).

janvorli commented:
@jkotas as for the new interface being created here and in @davmason's change, I suppose we will want to create two new versions, otherwise we would need to merge out changes into one, right?

jkotas commented:
> create two new version

I do not think that's required as long as we ship both changes in the same release cycle.

janvorli commented:
I can see that the Unix build is broken for the DAC build, I need to investigate it.

davmason commented:
> 
> 
> > create two new version
> 
> I do not think that's required as long as we ship both changes in the same release cycle.

Right, as long as it's the same release we can change the interfaces.

@janvorli it should be fairly easy to merge, whoever merges last can just add their methods to the existing interface in sospriv.idl and then regenerate sospriv.h and make sure the guid definition is right in sospriv_i.cpp. There shouldn't be any other work.

mikem8361 commented on src/coreclr/src/inc/clrprivbinderutil.h:
```diff
@@ -25,9 +25,7 @@
-interface ICLRPrivAssembly;
-typedef DPTR(ICLRPrivAssembly) PTR_ICLRPrivAssembly;
-typedef DPTR(ICLRPrivBinder) PTR_ICLRPrivBinder;
+//interface ICLRPrivAssembly;
```
Did you intentionally leave this commented out?

janvorli replied:
I should have removed that, I have missed it when looking through the code before commit.

jkotas commented on src/coreclr/src/binder/inc/assembly.hpp:
```diff
@@ -148,7 +153,7 @@ namespace BINDER_SPACE
         AssemblyName            *m_pAssemblyName;
```
This has a lot of non-DACized pointers. It is hard to see that none of them will get used by the DAC.

janvorli replied:
I have DACized only the ones that are needed for getting the ALC. It was painful enough that way.

jkotas replied:
I understand that it was painful. It is hard to see that it is correct. We prefer the type members to be fully DACized. Partially DACized types are bug farms.

mikem8361 reviewed (approved):
I do like this approach better.

jkotas commented on src/coreclr/src/vm/pefile.cpp:
```diff
@@ -2463,22 +2463,22 @@ PTR_ICLRPrivBinder PEFile::GetBindingContext()
-#ifndef DACCESS_COMPILE
-AssemblyLoadContext* PEFile::GetAssemblyLoadContext()
+PTR_AssemblyLoadContext PEFile::GetAssemblyLoadContext()
 {
```
Would it be better to store the `PTR_AssemblyLoadContext` and the object handle in `AssemblyLoadContext` base class to avoid DACizing the whole binder machinery?

jkotas replied:
I am worried that we will keep finding holes in the binder DACization.

janvorli replied:
> Would it be better to store the `PTR_AssemblyLoadContext` in the PEFile and the object handle in `AssemblyLoadContext` base class to avoid DACizing the whole binder machinery?

I will try to do it that way.

hoyosjs commented on src/coreclr/src/debug/daccess/request.cpp:
```diff
@@ -4500,3 +4500,29 @@ HRESULT ClrDataAccess::GetMethodsWithProfilerModifiedIL(CLRDATA_ADDRESS mod, CLR
+        return E_INVALIDARG;
+
+    SOSDacEnter();
+    PTR_MethodTable pMT = PTR_MethodTable(TO_TADDR(methodTable));
```
>TO_TADDR [](start = 42, length = 8)

Can you please change this to CLRDATA_ADDRESS_TO_TADDR

janvorli replied:
Sure. Btw, I am redoing it based on @jkotas feedback to simplify things considerably, so please wait with reviewing a bit.

janvorli committed:
Enhance DAC to allow getting AssemblyLoadContext for objects

This change adds support to DAC for getting AssemblyLoadContext
for a MethodTable. The returned AsseblyLoadContext is the context
into which the related type was loaded.

janvorli committed:
Change TO_TADDR to CLRDATA_ADDRESS_TO_TADDR

janvorli committed:
Fix Unix build

janvorli commented:
@jkotas I have updated the PR based on your suggestion of storing the AssemblyLoadContext in the PEFile. It worked quite well, the only little ugly thing being the need to set the ALC for the S.P.C. separately, as the PEFile (PEAssembly) for it is created during the ICLRRuntimeHost4::Start, but the related ALC much later during ICLRRuntimeHost4::CreateAppDomainWithManager.

jkotas commented on src/coreclr/src/vm/pefile.cpp:
```diff
@@ -1921,8 +1922,25 @@ PEAssembly::PEAssembly(
 #endif
-}
 
+#ifndef DACCESS_COMPILE
```
The whole function is `!DACCESS_COMPILE`. This ifdef should not be needed.

janvorli replied:
Strange, I have thought I have removed it already.

jkotas commented on src/coreclr/src/vm/appdomain.cpp:
```diff
@@ -4580,6 +4580,8 @@ IUnknown *AppDomain::CreateBinderContext()
         // Initialize the assembly binder for the default context loads for CoreCLR.
         IfFailThrow(CCoreCLRBinderHelper::DefaultBinderSetupContext(DefaultADID, &m_pTPABinderContext));
+
+        SystemDomain::SystemFile()->SetAssemblyLoadContext(m_pTPABinderContext);
```
Why do we need to set it explicitly? I would expect the constructor to set it for corelib too.

janvorli replied:
I've mentioned it in my comment above. the PEFile (PEAssembly) for it is created during the ICLRRuntimeHost4::Start, but the related ALC much later during ICLRRuntimeHost4::CreateAppDomainWithManager.

jkotas replied:
Maybe useful to add comment.

janvorli commented:
There are two tests failing. 
* System.Runtime.Loader.Tests.RefEmitLoadContextTests.LoadRefEmitAssembly
* System.Runtime.Loader.RefEmitLoadContext.Tests

It seems there is one more way how the `AssemblyLoadContext` would change after the PEFile creation. In `Assembly::CreateDynamic`, there is a call `pFile->SetFallbackLoadContextBinder(pFallbackLoadContextBinder);` and the `m_pFallbackContext` is used in the `AssemblyLoadContext` computation. I guess that's causing those two tests to fail. They check for `AssemblyLoadContext` type and get the default one instead of a specific one.

janvorli committed:
Fix dynamic assemblies case and reflect PR feedback

jkotas commented on src/coreclr/src/vm/pefile.cpp:
```diff
@@ -2331,6 +2335,24 @@ void PEFile::EnsureImageOpened()
         GetILimage()->GetLayout(PEImageLayout::LAYOUT_ANY,PEImage::LAYOUT_CREATEIFNEEDED)->Release();
 }
 
+AssemblyLoadContext* PEFile::LookupAssemblyLoadContext()
```
`SetAssemblyLoadContext` + `LookupAssemblyLoadContext` can be merged into a single method like `void SetupAssemblyLoadContext()` that computes it from the other PEFile state and sets it.

jkotas commented on src/coreclr/src/vm/pefile.h:
```diff
@@ -566,20 +572,34 @@ class PEFile
+    PTR_AssemblyLoadContext GetAssemblyLoadContext()
+    {
+        LIMITED_METHOD_CONTRACT;
+
```
Assert that `m_pAssemblyLoadContext` is non-NULL?

janvorli committed:
Cleanup - reflect PR feedback
